// @clawhub/cli/src/commands/login.ts
import { execSync } from "node:child_process";
import { saveToken } from "../auth.js";
import { SIGMA_AUTH_URL, SIGMA_CLIENT_ID } from "../config.js";
import { formatError, formatSuccess } from "../format.js";

function openBrowser(url: string): void {
  const cmd =
    process.platform === "darwin"
      ? "open"
      : process.platform === "win32"
        ? "start"
        : "xdg-open";

  try {
    execSync(`${cmd} "${url}"`, { stdio: "ignore" });
  } catch {
    console.log(`Open this URL in your browser:\n  ${url}`);
  }
}

interface DeviceResponse {
  device_code: string;
  user_code: string;
  verification_uri: string;
  verification_uri_complete?: string;
  expires_in: number;
  interval: number;
}

interface TokenResponse {
  access_token: string;
  token_type: string;
  expires_in?: number;
  error?: string;
  error_description?: string;
}

export async function login(): Promise<void> {
  console.log("Starting Sigma Auth device flow...\n");

  const deviceRes = await fetch(`${SIGMA_AUTH_URL}/api/auth/device/code`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      client_id: SIGMA_CLIENT_ID,
      scope: "openid profile",
    }),
  });

  if (!deviceRes.ok) {
    const text = await deviceRes.text();
    console.error(formatError(`Device authorization failed: ${text}`));
    process.exit(1);
  }

  const device: DeviceResponse = await deviceRes.json();
  const verifyUrl =
    device.verification_uri_complete ||
    `${device.verification_uri}?user_code=${device.user_code}`;

  console.log(`Go to: ${device.verification_uri}`);
  console.log(`Enter code: ${device.user_code}\n`);

  openBrowser(verifyUrl);

  console.log("Waiting for authorization...");

  let pollInterval = (device.interval || 5) * 1000;
  const deadline = Date.now() + device.expires_in * 1000;

  while (Date.now() < deadline) {
    await new Promise((r) => setTimeout(r, pollInterval));

    const tokenRes = await fetch(`${SIGMA_AUTH_URL}/api/auth/device/token`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        grant_type: "urn:ietf:params:oauth:grant-type:device_code",
        device_code: device.device_code,
        client_id: SIGMA_CLIENT_ID,
      }),
    });

    const tokenData: TokenResponse = await tokenRes.json();

    if (tokenData.access_token) {
      await saveToken(tokenData.access_token);
      console.log(formatSuccess("Logged in."));
      return;
    }

    if (tokenData.error === "authorization_pending") {
      continue;
    }

    if (tokenData.error === "slow_down") {
      pollInterval += 5000;
      continue;
    }

    if (tokenData.error === "expired_token") {
      console.error(formatError("Authorization expired. Please try again."));
      process.exit(1);
    }

    if (tokenData.error) {
      console.error(
        formatError(tokenData.error_description || tokenData.error)
      );
      process.exit(1);
    }
  }

  console.error(formatError("Authorization timed out. Please try again."));
  process.exit(1);
}
