// @clawhub/cli/src/commands/publish.ts
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { isAbsolute, join, relative } from "node:path";

import { parseSkillFrontmatter, validateSkillContent } from "@clawhub/sdk";
import { publishSkill } from "../api.js";
import { requireToken } from "../auth.js";
import { formatError, formatSuccess } from "../format.js";

const SKIP_DIRS = new Set(["node_modules", ".git", ".DS_Store", "__pycache__"]);
const BINARY_EXTS = new Set([
  ".png",
  ".jpg",
  ".jpeg",
  ".gif",
  ".ico",
  ".woff",
  ".woff2",
  ".ttf",
  ".eot",
  ".zip",
  ".tar",
  ".gz",
  ".exe",
  ".bin",
]);

function collectFiles(
  dir: string,
  base: string
): { path: string; content: string }[] {
  const files: { path: string; content: string }[] = [];

  for (const entry of readdirSync(dir)) {
    if (SKIP_DIRS.has(entry)) continue;
    if (entry.startsWith(".")) continue;

    const fullPath = join(dir, entry);
    const stat = statSync(fullPath);

    if (stat.isDirectory()) {
      files.push(...collectFiles(fullPath, base));
    } else if (stat.isFile()) {
      const ext = entry.slice(entry.lastIndexOf(".")).toLowerCase();
      if (BINARY_EXTS.has(ext)) continue;
      if (stat.size > 1024 * 1024) continue;

      const relPath = relative(base, fullPath);
      const content = readFileSync(fullPath, "utf8");
      files.push({ path: relPath, content });
    }
  }

  return files;
}

export async function publish(path?: string): Promise<void> {
  const token = await requireToken();
  const dir = path
    ? isAbsolute(path)
      ? path
      : join(process.cwd(), path)
    : process.cwd();

  if (!existsSync(dir) || !statSync(dir).isDirectory()) {
    console.error(formatError(`Not a directory: ${dir}`));
    process.exit(1);
  }

  const skillMdPath = existsSync(join(dir, "SKILL.md"))
    ? join(dir, "SKILL.md")
    : existsSync(join(dir, "skill.md"))
      ? join(dir, "skill.md")
      : null;

  if (!skillMdPath) {
    console.error(formatError("No SKILL.md found in directory."));
    process.exit(1);
  }

  const skillContent = readFileSync(skillMdPath, "utf8");
  const validation = validateSkillContent(skillContent);

  if (!validation.valid) {
    console.error(formatError("Invalid SKILL.md:"));
    for (const e of validation.errors) {
      console.error(`  - ${e}`);
    }
    process.exit(1);
  }

  const frontmatter = parseSkillFrontmatter(skillContent);
  if (!frontmatter.success) {
    console.error(
      formatError(`Failed to parse frontmatter: ${frontmatter.error}`)
    );
    process.exit(1);
  }

  const { name, description, version } = frontmatter.data;
  const slug = name;

  console.log(`Publishing ${slug}@${version}...`);

  const files = collectFiles(dir, dir);

  try {
    const result = await publishSkill(
      {
        slug,
        name,
        description,
        version,
        tags: frontmatter.data.tags,
        homepage: frontmatter.data.homepage,
        files,
      },
      token
    );
    console.log(formatSuccess(`Published ${result.slug}@${result.version}`));
  } catch (err) {
    console.error(
      formatError(err instanceof Error ? err.message : String(err))
    );
    process.exit(1);
  }
}
