// @clawhub/cli/src/api.ts
import { DEFAULT_REGISTRY } from "./config.js";

export class ApiError extends Error {
  constructor(
    public status: number,
    message: string
  ) {
    super(message);
    this.name = "ApiError";
  }
}

async function request(
  path: string,
  options: RequestInit = {},
  token?: string
): Promise<unknown> {
  const url = `${DEFAULT_REGISTRY}/api/v1${path}`;
  const headers: Record<string, string> = {
    ...(options.headers as Record<string, string>),
  };

  if (token) {
    headers.Authorization = `Bearer ${token}`;
  }

  if (options.body && !headers["Content-Type"]) {
    headers["Content-Type"] = "application/json";
  }

  const res = await fetch(url, { ...options, headers });

  if (!res.ok) {
    if (res.status === 401) {
      throw new ApiError(401, 'Not authenticated. Run "clawnet login" first.');
    }
    const body = await res.json().catch(() => ({ error: res.statusText }));
    throw new ApiError(
      res.status,
      (body as { error?: string }).error || res.statusText
    );
  }

  return res.json();
}

export interface SkillSummary {
  slug: string;
  name: string;
  description: string;
  authorBapId: string;
  latestVersion: string;
  latestTxId: string;
  homepage?: string;
  tags?: string[];
  starCount: number;
  downloadCount: number;
  createdAt: number;
  updatedAt: number;
}

export interface SkillDetail {
  skill: SkillSummary & { deleted: boolean; downloadCountAllTime: number };
  latestVersion: {
    version: string;
    txId: string;
    content: string;
    contentType: string;
    changelog?: string;
    files: { path: string; content: string }[];
    publishedAt: number;
    onChain: boolean;
  } | null;
  author: { bapId: string; pubkey: string } | null;
}

export interface SearchResult {
  slug: string;
  name: string;
  description: string;
  authorBapId: string;
  latestVersion: string;
  score: number;
}

export interface WhoamiResponse {
  user: { bapId: string; pubkey: string };
}

export async function listSkills(
  sort = "updated",
  limit = 50,
  cursor?: string
): Promise<{ skills: SkillSummary[]; hasMore: boolean; cursor?: string }> {
  const params = new URLSearchParams({ sort, limit: String(limit) });
  if (cursor) params.set("cursor", cursor);
  return request(`/skills?${params}`) as Promise<{
    skills: SkillSummary[];
    hasMore: boolean;
    cursor?: string;
  }>;
}

export async function getSkill(slug: string): Promise<SkillDetail> {
  return request(`/skills/${encodeURIComponent(slug)}`) as Promise<SkillDetail>;
}

export async function searchSkills(
  q: string,
  limit = 20
): Promise<{ results: SearchResult[] }> {
  const params = new URLSearchParams({ q, limit: String(limit) });
  return request(`/search?${params}`) as Promise<{
    results: SearchResult[];
  }>;
}

export async function publishSkill(
  payload: {
    slug: string;
    name: string;
    description: string;
    version: string;
    changelog?: string;
    tags?: string[];
    homepage?: string;
    files: { path: string; content: string }[];
  },
  token: string
): Promise<{ ok: boolean; slug: string; version: string }> {
  return request(
    "/skills",
    { method: "POST", body: JSON.stringify(payload) },
    token
  ) as Promise<{ ok: boolean; slug: string; version: string }>;
}

export async function fetchWhoami(token: string): Promise<WhoamiResponse> {
  return request("/whoami", {}, token) as Promise<WhoamiResponse>;
}

export async function starSkill(
  slug: string,
  token: string
): Promise<{ ok: boolean }> {
  return request(
    `/stars/${encodeURIComponent(slug)}`,
    { method: "POST" },
    token
  ) as Promise<{ ok: boolean }>;
}

export async function unstarSkill(
  slug: string,
  token: string
): Promise<{ ok: boolean }> {
  return request(
    `/stars/${encodeURIComponent(slug)}`,
    { method: "DELETE" },
    token
  ) as Promise<{ ok: boolean }>;
}
